source/api_c/tests/test_deeppot_a.cc

```diff
@@ -0,0 +1,116 @@
+  };
+  int natoms;
+  double expected_tot_e;
+  std::vector<double>expected_tot_v;
+
+  DP_DeepPot* dp; 
+
+  void SetUp() override {
+    const char* file_name = "../../tests/infer/deeppot.pbtxt";
+    const char* model_file = "deeppot.pb";
+    DP_ConvertPbtxtToPb(file_name, model_file);
+
+    dp = DP_NewDeepPot(model_file);
+
+    natoms = expected_e.size();
+    EXPECT_EQ(natoms * 3, expected_f.size());
+    EXPECT_EQ(natoms * 9, expected_v.size());
+    expected_tot_e = 0.;
+    expected_tot_v.resize(9);
+    std::fill(expected_tot_v.begin(), expected_tot_v.end(), 0.);
+    for(int ii = 0; ii < natoms; ++ii){
+      expected_tot_e += expected_e[ii];
+    }
+    for(int ii = 0; ii < natoms; ++ii){
+      for(int dd = 0; dd < 9; ++dd){
+	      expected_tot_v[dd] += expected_v[ii*9+dd];
+      }
+    }
+  };
+
+  void TearDown() override {
+    remove( "deeppot.pb" ) ;
+  };
+};
+
+TEST_F(TestInferDeepPotA, double_infer)
+{
+  double* ener_ = new double;
+  double* force_ = new double[natoms * 3];
+  double* virial_ = new double[9];
+  DP_DeepPotCompute (dp, natoms, coord, atype, box, ener_, force_, virial_);
+
+  double ener = *ener_;
+  std::vector<double> force(force_, force_ + natoms * 3);
+  std::vector<double> virial(virial_, virial_ + 9);
+
+  EXPECT_LT(fabs(ener - expected_tot_e), 1e-10);
+  for(int ii = 0; ii < natoms*3; ++ii){
+    EXPECT_LT(fabs(force[ii] - expected_f[ii]), 1e-10);    
+  }
+  for(int ii = 0; ii < 3*3; ++ii){
+    EXPECT_LT(fabs(virial[ii] - expected_tot_v[ii]), 1e-10);
+  }
+}
+
+TEST_F(TestInferDeepPotA, float_infer)
+{
+  double* ener_ = new double;
+  float* force_ = new float[natoms * 3];
+  float* virial_ = new float[9];
+
+  DP_DeepPotComputef (dp, natoms, coordf, atype, boxf, ener_, force_, virial_);
+
+  double ener = *ener_;
+  std::vector<float> force(force_, force_ + natoms * 3);
+  std::vector<float> virial(virial_, virial_ + 9);
+
+  EXPECT_LT(fabs(ener - expected_tot_e), 1e-6);
+  for(int ii = 0; ii < natoms*3; ++ii){
+    EXPECT_LT(fabs(force[ii] - expected_f[ii]), 1e-6);    
```

Comment:
Im surprised that the expected precision of the C interface is higher than that of the C++ interface (1e-4)

Reply:
The atomic energy does not have 1e-6 precision.